opencog/cogserver/server/AgentRunnerBase.cc

```diff
@@ -0,0 +1,165 @@
+ * published by the Free Software Foundation and including the exceptions
+ * at http://opencog.org/wiki/Licenses
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program; if not, write to:
+ * Free Software Foundation, Inc.,
+ * 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+ */
+
+#include <algorithm>
+#include <chrono>
+#include <opencog/cogserver/server/AgentRunnerBase.h>
+#include <opencog/cogserver/server/CogServer.h>
+#include <opencog/util/Logger.h>
+#include <opencog/util/platform.h>
+
+using namespace std;
+using namespace chrono;
+
+
+namespace opencog
+{
+
+struct equal_to_id: public std::binary_function<AgentPtr, const std::string&, bool>
+{
+        bool operator()(AgentPtr a, const std::string& cid) const
+        {
+            return (a->classinfo().id != cid);
+        }
+};
+
+AgentRunnerBase::AgentRunnerBase(std::string runner_name): name(runner_name),
+        cycleCount(1)
+{
+}
+
+AgentRunnerBase::~AgentRunnerBase()
+{
+}
+
+void AgentRunnerBase::setName(std::string new_name)
+{
+    name = new_name;
+}
+
+const std::string& AgentRunnerBase::getName() const
+{
+    return name;
+}
+
+unsigned long AgentRunnerBase::getCycleCount() const
+{
+    return cycleCount;
+}
+
+void AgentRunnerBase::addAgent(AgentPtr a)
+{
+    agents.push_back(a);
+}
+
+void AgentRunnerBase::removeAgent(AgentPtr a)
+{
+    AgentSeq::iterator ai = std::find(agents.begin(), agents.end(), a);
+    if (ai != agents.end()) {
+        agents.erase(ai);
+        cogserver().systemActivityTable().clearActivity(a);
+        a->stop();
+        logger().debug("[CogServer::%s] stopped agent \"%s\"", name.c_str(),
+            a->to_string().c_str());
+    }
+}
+
+void AgentRunnerBase::removeAllAgents(const std::string& id)
+{
+    // place agents with classinfo().id == id at the end of the container
+    AgentSeq::iterator last =
+        std::partition(agents.begin(), agents.end(),
+                       boost::bind(equal_to_id(), _1, id));
+
+    std::for_each(last, agents.end(), [] (AgentPtr &a) {
+        cogserver().systemActivityTable().clearActivity(a);
+        a->stop();
+    });
+
+    // remove those agents from the main container
+    agents.erase(last, agents.end());
+
+    logger().debug("[CogServer::%s] stopped all agents of type \"%s\"",
+        name.c_str(), id.c_str());
+}
+
+void AgentRunnerBase::removeAllAgents()
+{
+    for (auto &a: agents)
+    {
+        cogserver().systemActivityTable().clearActivity(a);
+        a->stop();
+    }
+    agents.clear();
+    logger().debug("[CogServer::%s] stopped all agents", name.c_str());
+}
+
+void AgentRunnerBase::runAgent(AgentPtr a)
```

Comment:
Please give this method a better name. You use the string runAgent here, and also in runAgentsMutex and also in a boolean called runAgents, and also in a condition called runAgentsCond.   That means a grep for the string "runAgent" has 27 hits, so far, in this patch. That makes it extremely difficult to trace through the code, because of all of the occurrences of the string "runAgent", only 1/4th of them actually refer to this method -- the rest refer to other things.   

alternately you could rename the mutexes the boolean and the contiditon variable -- they don't really have anything to do with this method, and so should not have conflicting names.